cpp/src/arrow/array/validate.cc

```diff
@@ -166,6 +166,80 @@ struct ValidateArrayImpl {
     return Status::OK();
   }
 
+  Status Visit(const Date64Type& type) {
+    RETURN_NOT_OK(ValidateFixedWidthBuffers());
+
+    using c_type = typename Date64Type::c_type;
+    if (full_validation) {
+      constexpr c_type kFullDay = 1000 * 60 * 60 * 24;
+      return VisitArrayDataInline<Date64Type>(
+          data,
+          [&](c_type date) {
+            if (date % kFullDay != 0) {
+              return Status::Invalid(type, date,
+                                     "ms does not represent a whole number of days");
+            }
+            return Status::OK();
+          },
+          []() { return Status::OK(); });
+    }
+    return Status::OK();
+  }
+
+  Status Visit(const Time32Type& type) {
+    RETURN_NOT_OK(ValidateFixedWidthBuffers());
+
+    using c_type = typename Time32Type::c_type;
+    if (full_validation) {
+      constexpr c_type kFullDaySeconds = 60 * 60 * 24;
+      constexpr c_type kFullDayMillis = kFullDaySeconds * 1000;
+      return VisitArrayDataInline<Time32Type>(
+          data,
+          [&](c_type time) {
+            if (type.unit() == TimeUnit::SECOND && (time < 0 || time >= kFullDaySeconds)) {
+              return Status::Invalid(type, " ", time,
+                                     "s does not fit within the acceptable range of ",
```

Comment:
_"does not fit within"_ --> _"is not within"_ I think this message represents better the invalid value. 
Same applies for the other cases below.